Add new entry: working with JSON in Go (#5906)
Codecademy/docs: content/go/concepts/working-with-json/working-with-json.md

---
Title: 'Working with JSON'

> Description: 'Working with JSON files, requests, or responses is an essential part of interacting with data that will be used to exchange information between the Frontend and Backend in web applications.'

Reviewer: Avoid using "will be" as it implies futurity unnecessarily.

```suggestion
Description: 'Working with JSON files, requests, or responses is essential to interacting with data that is exchanged between the Frontend and Backend in web applications.'
```

Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:

>   - 'JSON'
>   - 'Marshal'
>   - 'Unmarshal'
>   - 'Encode'
>   - 'Decode'

Reviewer: These are not valid tags, please check [tags.md](https://github.com/Codecademy/docs/blob/main/documentation/tags.md) for more reference. You can use tags from `tags.md` as follows:

```suggestion
  - 'JSON'
  - 'Data'
  - 'Development'
  - 'Objects'
```

CatalogContent:
  - 'learn-go'
  - 'paths/back-end-engineer-career-path'
---

> **JSON** is one of the most commonly used data interchange formats for communication between web applications, other services, mobile applications, and so on.
> In Go there are functions dedicated to work with **JSON**, here we are going to describe the most used.

Reviewer: We can rewrite this for simplicity like this:

```suggestion
**JSON** is one of the most commonly used data exchange formats for communication between web applications, other services, mobile applications, etc.

Go provides built-in functions to work with JSON effectively. Below are some of the most commonly used functions:
```

> 

Reviewer: The table needs some introductory text:

```suggestion
Below is a comparison between `json.Marshal` and `json.NewDecoder` based on their use cases and performance considerations.
```

Reviewer: Ideally, this table should come at last, after discussing the syntax and example use case.

> | Feature      | json.Marshal                                                                  | json.NewDecoder                                  |

Reviewer: Enclose code-related keywords in backticks.

```suggestion
| Feature      | `json.Marshal`                                                                  | `json.NewDecoder`                                  |
```

| ------------ | ----------------------------------------------------------------------------- | ------------------------------------------------ |

> | Input        | Go value (struct, map, etc.)                                                  | io.ReadCloser (network connection)               |

Reviewer: Enclose code-related keywords in backticks.

```suggestion
| Input        | Go value (`struct`, `map`, etc.)                                                  | `io.ReadCloser` (network connection)               |
```

| Output       | JSON byte slice                                                               | Go value (struct, map, etc.)                     |
| Memory Usage | Can potentially use more memory if the input data is large                    | Generally more memory-efficient for large inputs |
| Use Cases    | Converting Go data to JSON for various purposes, suitable for small data sets | Decoding JSON data from streams or large files   |

> ## Marshal and Unmarshal

Reviewer: Use consistent verb forms in section headers.

```suggestion
## Using Marshal and Unmarshal
```

> In Go `json.Unmarshall` is used to decode JSON data formatted as a string. This allows us to add new elements, extract specific values like configurations, and modify existing ones. Subsequently, Next is `json.Marshal` a function that is used to encode the modified Go data structure into a string, which can then be saved to a file or used for other purposes, as demonstrated in the example below:

Reviewer: We should rewrite this for simplicity, keeping `json.Marshal` first:

```suggestion

In Go, `json.Marshal` encodes a Go data structure into a JSON-formatted string. This encoded data can then be saved to a file, transmitted, or used in other applications. 

Conversely, `json.Unmarshal` decodes JSON data into a Go data structure. This allows for operations such as adding new elements, extracting specific values (e.g., configurations), or modifying existing ones, as demonstrated in the example below:
```

}

type Meals struct {
    Meals []Meal `json:"meals"`
}

func main() {
    menu := `{
        "meals": [
          {"name": "Pizza Margherita", "ingredients": ["Dough", "Tomato Sauce", "Mozzarella Cheese", "Tomato", "Basil"]},
          {"name": "Spaghetti Carbonara", "ingredients": ["Spaghetti", "Eggs", "Pancetta", "Pecorino Romano Cheese", "Black Pepper"]},
          {"name": "Chicken Stir-fry", "ingredients": ["Chicken", "Vegetables (e.g., Broccoli, Carrots, Onions, Peppers)", "Rice", "Soy Sauce", "Ginger", "Garlic"]},
          {"name": "Tacos al Pastor", "ingredients": ["Pork", "Pineapple", "Tortillas", "Onion", "Cilantro"]},
          {"name": "Caesar Salad", "ingredients": ["Romaine Lettuce", "Croutons", "Parmesan Cheese", "Caesar Dressing"]}
        ]
    }`

    // We need a variable to store the data decoded from the unmarshaling process.
    meals := Meals{}

    // To unmarshal the data, we need to provide:
    // 1. Data in the format of a slice of bits.
    // 2. A pointer to the location where the data will be stored.
    if err := json.Unmarshal([]byte(menu), &meals); err != nil {
        // Handle the errors in case they occur.
        fmt.Println(err)
    }

>     // In case we need to add a new object to the

Reviewer: This comment is incomplete.

```suggestion
    // To add a new object to the menu, create an instance of Meal and append it to the meals.Meals slice.
```

    meal := Meal{
        Name:        "Pupusas",
        Ingredients: []string{"Cheese", "Refried beans", "Tomato Sauce", "Rice flour", "Pickled cabbage"},
    }

    // Add new meal for the menu
    meals.Meals = append(meals.Meals, meal)

    // Add new meal in line for the menu
    meals.Meals = append(meals.Meals, Meal{Name: "Sushi", Ingredients: []string{"Rice", "Fish", "Seaweed", "Wasabi", "Soy Sauce", "Ginger"}})

    // List the menu of meals
    for _, meal := range meals.Meals {
        fmt.Printf("---\nMeal: %s\ningredients: %s \n", meal.Name, strings.Join(meal.Ingredients, ", "))
    }

    // Serialize the object to a slice of bytes
    data, err := json.Marshal(meals)
    if err != nil {
        fmt.Println(err)
    }
    fmt.Println("-------------")
    fmt.Println(string(data))

    // Crate an empty file
    file, err := os.Create("meals.json")
    if err != nil {
        fmt.Println("Error opening file:", err)
        return
    }

    defer file.Close()

    // Write data into the file previously created
    if _, err = file.Write(data); err != nil {
        fmt.Println(err)
        return
    }

}
```

> ## NewEncoder and NewDecoder

Reviewer: Use consistent verb forms in section headers.

```suggestion
## Using NewEncoder and NewDecoder
```

> In Go `json.NewDecoder` is used to format data received from a REST API. Other function that is commonly used is `json.NewEncoder` to exchange data between backend side and external clients that require a JSON response.

Reviewer: This should be rewritten such that `json.NewEncoder` comes first:

```suggestion
In Go, `json.NewEncoder` is commonly used to encode data into a JSON format, making it suitable for exchanging information between the backend and external clients that require a JSON response. On the other hand, `json.NewDecoder` is used to decode data received from a REST API, transforming it into a usable Go data structure.
```

In the following example, we will demonstrate the usage of both functions. We will make a call to an external REST API and enhance the result by adding products made from the fruit specified in the query parameter. For example: `http://localhost:4444/fruit?name=Strawberry`

```go
package main

import (
    "encoding/json"
    "errors"
    "fmt"
    "io"
    "log"
    "net/http"
    "strings"
)

type ProductsByFruit map[string][]string

type Fruit struct {
    Name     string   `json:"name"`
    ID       int      `json:"id"`
    Family   string   `json:"family"`
    Order    string   `json:"order"`
    Genus    string   `json:"genus"`
    Products []string `json:"products"`
}

func main() {
    mux := http.NewServeMux()

>     mux.HandleFunc("POST /fruit", func(w http.ResponseWriter, r *http.Request) {

Reviewer: The `mux.HandleFunc` method should not include the HTTP method (`POST`) in the path. The path should just specify the URL pattern, and the HTTP method should be handled separately using `http.MethodPost`.

Fix:

```go
mux.HandleFunc("/fruit", func(w http.ResponseWriter, r *http.Request) {
    if r.Method != http.MethodPost {
        http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
        return
    }
    // Continue handling the request...
})
```

        // Get query params from URL
        params := r.URL.Query()

        // We are going to validate the name of the fruit
        if fruitName := params.Get("name"); len(fruitName) > 0 {

            // Set the specific content type for the response
            w.Header().Set("Content-Type", "application/json")

>             // Get the results of the invocation to the REST API.
>             result, err := GetFruitData(fruitName)
> 
>             // We need to handle any errors that occur during the decoding process.
>             if err != nil {
>                 http.Error(w, "Error fruit data not found", http.StatusNotFound)
>                 return
>             }

Reviewer: Avoid ignoring errors directly. For example, in `GetFruitData`, checking `response.StatusCode` before handling the `response.Body` could lead to confusion if `response` is `nil`.

```suggestion
            // Get the results of the invocation to the REST API.
            result, err := GetFruitData(fruitName)

            // We need to handle any errors that occur during the decoding process.
            if err != nil {
            	return nil, fmt.Errorf("error fetching data: %w", err)
	    }
```

            // We need a variable to store the results from the decoding process.
            fruit := Fruit{}

            // To Decode the data, we need to provide:
            // 1. Data in the format of io.ReadCloser.
            // 2. A pointer to the location where the data will be stored.

>             err = json.NewDecoder(result).Decode(&fruit)

Reviewer: The `GetFruitData` function doesn't check if the response is `nil` before using it. In the case of an error (e.g., network failure), response could be `nil`, leading to a runtime panic.

Fix:

```go
if response == nil {
    return nil, errors.New("no response from the server")
}
```

---
Title: 'Working with JSON'
Description: 'Working with JSON files, requests, or responses is essential to interacting with data that is exchanged between the Frontend and Backend in web applications.'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'JSON'
  - 'Data'
  - 'Development'
  - 'Objects'
CatalogContent:
  - 'learn-go'
  - 'paths/back-end-engineer-career-path'
---

> **JSON** is one of the most commonly used data exchange formats for communication between web applications, other services, mobile applications, etc.
> 
> Go provides built-in functions to work with JSON effectively.

Reviewer 2: ```suggestion
**JSON** is one of the most commonly used data exchange formats for communication between web applications, other services, mobile applications, etc. Go provides built-in functions to work with JSON effectively.
```

    }

    // Add the new meal to the list.
    meals.Meals = append(meals.Meals, meal)

    // Add another meal inline.
    meals.Meals = append(meals.Meals, Meal{Name: "Sushi", Ingredients: []string{"Rice", "Fish", "Seaweed", "Wasabi", "Soy Sauce", "Ginger"}})

    // List the menu of meals.
    for _, meal := range meals.Meals {
        fmt.Printf("---\nMeal: %s\nIngredients: %s\n", meal.Name, strings.Join(meal.Ingredients, ", "))
    }

    // Serialize the object to a slice of bytes.
    data, err := json.Marshal(meals)
    if err != nil {
        fmt.Println("Error marshalling data:", err)
        return
    }

    fmt.Println("-------------")
    fmt.Println(string(data))

    // Create an empty file.
    file, err := os.Create("meals.json")
    if err != nil {
        fmt.Println("Error creating file:", err)
        return
    }

    // Ensure the file is closed after the program finishes.
    defer file.Close()

    // Write the serialized data into the file.
    if _, err = file.Write(data); err != nil {
        fmt.Println("Error writing to file:", err)
        return
    }

    fmt.Println("Data has been written to meals.json successfully.")
}
```

> When you run the above Go program, the following will happen:

Reviewer 2: ```suggestion
When running the above Go program, the following will happen:
```
